File: iphone/roadmap_ssl.m
/* roadmap_ssl.c - SSL connection for iPhone
 *
 * LICENSE:
 *
 *   Copyright 2011, Avi R.
 *   Copyright 2011, Waze Ltd
 *
 *   This file is part of RoadMap.
 *
 *   RoadMap is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   RoadMap is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with RoadMap; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */


#include <stdlib.h>

#include "roadmap_main.h"
#include "roadmap_iphonemain.h"
#include "roadmap_net.h"
#include "roadmap_ssl.h"

#import <CFNetwork/CFProxySupport.h>


struct roadmap_ssl_io_t {
   RoadMapSocket  s;
   void           *data;
   CFReadStreamRef readStream;
   CFWriteStreamRef writeStream;
   RoadMapNetSslConnectCallback onOpenCallback;
   RoadMapInput callback;
};


//////////////////////////////////////////////////////////////////////////////////////////////////
//void roadmap_ssl_set_secured_input (RoadMapIO *io, RoadMapInput callback, void *ssl_context) {
//   RoadMapSslIO ssl_io;
//   ssl_io->callback = callback;
//}

//////////////////////////////////////////////////////////////////////////////////////////////////
int roadmap_ssl_read (void *context, void *buffer, int buffer_size) {
  RoadMapSslIO ssl_io = (RoadMapSslIO) context;
   int bytes_read = CFReadStreamRead (ssl_io->readStream, buffer, buffer_size);
   
   return bytes_read;
}

//////////////////////////////////////////////////////////////////////////////////////////////////
static void read_stream_client_cb (CFReadStreamRef stream,
                                   CFStreamEventType eventType,
                                   void *clientCallBackInfo)
{
   //printf("read_stream_client_cb status: %d\n", CFReadStreamGetStatus(stream));
   if (eventType & kCFStreamEventOpenCompleted) {
      //CFStreamClientContext *context = (CFStreamClientContext*) clientCallBackInfo;
      //struct roadmap_main_ssl_io *io = clientCallBackInfo;//context->info;
      //NSLog(@"read_stream_client_cb: kCFStreamEventOpenCompleted");
      //      on_ssl_open_callback(io->s, io->data, io);
      //      free(io);
      //      return;
   }
   if (eventType & kCFStreamEventErrorOccurred) {
      CFErrorRef err = CFReadStreamCopyError(stream);
      if (err) {
         CFStringRef str = CFErrorCopyDescription(err);
         char buf[1024];
         
         if (str) {
            CFStringGetCString(str, buf, 1024, kCFStringEncodingUTF8);
            roadmap_log(ROADMAP_ERROR, "read_stream_client_cb: kCFStreamEventErrorOccurred description: '%s'", buf);
            CFRelease(str);
         }
         
         str = CFErrorCopyFailureReason (err);
         if (str) {
            CFStringGetCString(str, buf, 1024, kCFStringEncodingUTF8);
            roadmap_log(ROADMAP_ERROR, "read_stream_client_cb: kCFStreamEventErrorOccurred reason: '%s'", buf);
            CFRelease(str);
         }
         
         str = CFErrorCopyRecoverySuggestion (err);
         if (str) {
            CFStringGetCString(str, buf, 1024, kCFStringEncodingUTF8);
            roadmap_log(ROADMAP_ERROR, "read_stream_client_cb: kCFStreamEventErrorOccurred recovery: '%s'", buf);
            CFRelease(str);
         }
         
         CFRelease(err);
      }
      
      RoadMapSslIO io = clientCallBackInfo;//context->info;
      io->onOpenCallback(io->s, io->data, io, err_net_failed);
      roadmap_ssl_close(io);
      
      return;
   }
   
   if (eventType & kCFStreamEventEndEncountered) {
      //NSLog(@"read_stream_client_cb: kCFStreamEventEndEncountered");
      return;
   }
   
   if (eventType & kCFStreamEventHasBytesAvailable) {
      //NSLog(@"read_stream_client_cb: kCFStreamEventHasBytesAvailable");
      //      
      //UInt8 buffer[512];
      //            int bytes_read = CFReadStreamRead (stream, buffer, sizeof(buffer));
      //            if (bytes_read == -1)
      //               return;
      //      struct roadmap_main_ssl_io *io = clientCallBackInfo;//context->info;
      //on_ssl_read_callback(io->s, io->data, buffer, bytes_read);
      // io->callback();
      //      
      //      err = AudioFileStreamParseBytes (af_parser, bytes_read, buffer, parseFlag);
      //      if (err)
      //         //NSLog(@"error AudioFileStreamParseBytes");
      return;
   }
}

//////////////////////////////////////////////////////////////////////////////////////////////////
static void write_stream_client_cb (CFWriteStreamRef stream,
                                    CFStreamEventType eventType,
                                    void *clientCallBackInfo)
{
   //printf("write_stream_client_cb status: %d\n", CFReadStreamGetStatus(stream));
   if (eventType & kCFStreamEventOpenCompleted) {
      //NSLog(@"write_stream_client_cb: kCFStreamEventOpenCompleted");
      return;
   }
   if (eventType & kCFStreamEventErrorOccurred) {
      CFErrorRef err = CFWriteStreamCopyError(stream);
      if (err) {
         CFStringRef str = CFErrorCopyDescription(err);
         char buf[1024];
         
         if (str) {
            CFStringGetCString(str, buf, 1024, kCFStringEncodingUTF8);
            roadmap_log(ROADMAP_ERROR, "write_stream_client_cb: kCFStreamEventErrorOccurred description: '%s'", buf);
            CFRelease(str);
         }
         
         str = CFErrorCopyFailureReason (err);
         if (str) {
            CFStringGetCString(str, buf, 1024, kCFStringEncodingUTF8);
            roadmap_log(ROADMAP_ERROR, "write_stream_client_cb: kCFStreamEventErrorOccurred reason: '%s'", buf);
            CFRelease(str);
         }
         
         str = CFErrorCopyRecoverySuggestion (err);
         if (str) {
            CFStringGetCString(str, buf, 1024, kCFStringEncodingUTF8);
            roadmap_log(ROADMAP_ERROR, "write_stream_client_cb: kCFStreamEventErrorOccurred recovery: '%s'", buf);
            CFRelease(str);
         }
         
         CFRelease(err);
      }
      
      RoadMapSslIO io = clientCallBackInfo;//context->info;
      io->onOpenCallback(io->s, io->data, io, err_net_failed);
      roadmap_ssl_close(io);
      
      return;
   }
   
   if (eventType & kCFStreamEventEndEncountered) {
      //NSLog(@"write_stream_client_cb: kCFStreamEventEndEncountered");
      return;
   }
   
   if (eventType & kCFStreamEventCanAcceptBytes) {
      CFWriteStreamSetClient(stream, 0, NULL, NULL);
      //NSLog(@"write_stream_client_cb: kCFStreamEventCanAcceptBytes");
      RoadMapSslIO io = clientCallBackInfo;//context->info;
      io->onOpenCallback(io->s, io->data, io, succeeded);
      return;
   }
}

//////////////////////////////////////////////////////////////////////////////////////////////////
int roadmap_ssl_send (RoadMapSslIO io, const void *data, int length) {
   int res = CFWriteStreamWrite(io->writeStream, data, length);
   //printf("res: %d\n", res);
   if (res < 0) {
      CFErrorRef err = CFWriteStreamCopyError(io->writeStream);
      if (err) {
         CFStringRef str = CFErrorCopyDescription(err);
         char buf[1024];
         CFStringGetCString(str, buf, 1024, kCFStringEncodingUTF8);
         roadmap_log(ROADMAP_ERROR, "Error writing to SSL stream: '%s'", buf);
         CFRelease(str);
         CFRelease(err);
      }
   }
   
   return res;
}

//////////////////////////////////////////////////////////////////////////////////////////////////
int roadmap_ssl_open (RoadMapSocket s, void *data, RoadMapNetSslConnectCallback callback) {
   RoadMapSslIO ssl_io;
   CFReadStreamRef readStream;
   CFWriteStreamRef writeStream;
   CFStreamCreatePairWithSocket(NULL, roadmap_net_get_fd(s), &readStream, &writeStream);
   if (!readStream || !writeStream) {
      roadmap_log(ROADMAP_ERROR, "CFStreamCreatePairWithSocket() failed");
      return 0;
   }
   //printf("open_ssl status: %d\n", CFReadStreamGetStatus(readStream));
   
   ssl_io = malloc(sizeof (*ssl_io));
   ssl_io->s = s;
   ssl_io->data = data;
   ssl_io->readStream = readStream;
   ssl_io->writeStream = writeStream;
   ssl_io->onOpenCallback = callback;
   CFStreamClientContext context = {0, ssl_io, NULL, NULL, NULL};
   CFReadStreamSetClient(readStream,
                         kCFStreamEventHasBytesAvailable | kCFStreamEventErrorOccurred | kCFStreamEventEndEncountered
                         |kCFStreamEventOpenCompleted,
                         read_stream_client_cb,
                         &context);
   CFReadStreamScheduleWithRunLoop(readStream, CFRunLoopGetCurrent(), kCFRunLoopCommonModes);
   CFWriteStreamSetClient(writeStream,
                          kCFStreamEventCanAcceptBytes | kCFStreamEventErrorOccurred | kCFStreamEventEndEncountered
                          |kCFStreamEventOpenCompleted,
                          write_stream_client_cb,
                          &context);
   CFWriteStreamScheduleWithRunLoop(writeStream, CFRunLoopGetCurrent(), kCFRunLoopCommonModes);
   
   //Open read stream
   NSInputStream *inputStream = (NSInputStream *)readStream;
   [inputStream setProperty:NSStreamSocketSecurityLevelNegotiatedSSL forKey:NSStreamSocketSecurityLevelKey];
   
   if (!CFReadStreamOpen(readStream)) {
      roadmap_log(ROADMAP_ERROR, "CFReadStreamOpen() failed");
      return 0;
   }
   //printf("open_ssl status: %d\n", CFReadStreamGetStatus(readStream));
   
   //Open write stream
   NSOutputStream *outputStream = (NSOutputStream *)writeStream;
   [outputStream setProperty:NSStreamSocketSecurityLevelNegotiatedSSL forKey:NSStreamSocketSecurityLevelKey];
   
   if (!CFWriteStreamOpen(writeStream)) {
      roadmap_log(ROADMAP_ERROR, "CFWriteStreamOpen() failed");
      return 0;
   }
   //printf("open_ssl status: %d\n", CFWriteStreamGetStatus(writeStream));
   
   return 1;
}

//////////////////////////////////////////////////////////////////////////////////////////////////
void roadmap_ssl_close (void *context) {
   RoadMapSslIO ssl_io = (RoadMapSslIO) context;
   
   if (!context) {
      roadmap_log(ROADMAP_WARNING, "roadmap_ssl_close() - NULL context");
      return;
   }
   
   CFWriteStreamUnscheduleFromRunLoop(ssl_io->writeStream, CFRunLoopGetCurrent(), kCFRunLoopCommonModes);
   CFWriteStreamClose (ssl_io->writeStream);
   CFRelease(ssl_io->writeStream);
   
   CFReadStreamUnscheduleFromRunLoop(ssl_io->readStream, CFRunLoopGetCurrent(), kCFRunLoopCommonModes);
   CFReadStreamClose (ssl_io->readStream);
   CFRelease(ssl_io->readStream);
   
   free (ssl_io);
}

File: unix/roadmap_ssl.c
/* roadmap_ssl.c - Open SSL integration
 *
 *
 * LICENSE:
 *   Copyright 2011, Waze Ltd      Alex Agranovich (AGA)
 *
 *
 *
 *   This file is part of RoadMap.
 *
 *   RoadMap is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License V2 as published by
 *   the Free Software Foundation.
 *
 *   RoadMap is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with RoadMap; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * SYNOPSYS:
 *
 *   See roadmap_ssl.h
 *       roadmap_net.c
 *
 */

#include "roadmap.h"
#include "roadmap_ssl.h"
#include "roadmap_net.h"
#include <stdio.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <openssl/evp.h>

//======================== Local defines ========================

//======================== Local types ========================

struct roadmap_ssl_io_t {
   RoadMapSocket  s;
   void           *data;
   SSL_CTX *ctx;
   SSL *ssl;
   RoadMapNetSslConnectCallback on_connect; // TODO: Necessary ???
};

//======================== Local Declarations ========================

static const char* last_err_string( void );
static void log_ssl( void );

//======================== Globals ========================


/*
 ******************************************************************************
 */
int roadmap_ssl_open ( RoadMapSocket s, void *data, RoadMapNetSslConnectCallback callback )
{
   static BOOL initialized = FALSE;
   const SSL_METHOD *method;
   SSL_CTX *ctx;
   SSL *ssl;

   // Initialize the library
   if ( !initialized )
   {
      SSL_library_init();
   }
   /*
    * Initialize the ssl context
    */
   OpenSSL_add_all_algorithms();      /* Load cryptos, et.al. */
   SSL_load_error_strings();          /* Bring in and register error messages */
   method = SSLv23_client_method();    /* Create new client-method instance */
   ctx = SSL_CTX_new( method );       /* Create new context */
   if ( ctx == NULL )
   {
      roadmap_log( ROADMAP_ERROR, "Error obtaining ssl context: %s !!", last_err_string() );
      return 0;
   }

   ssl = SSL_new( ctx );                /* SSL connection state */
   SSL_set_fd( ssl, roadmap_net_get_fd( s ) );             /* Attach the socket */

   /*
    * Initialize the io context
    */

   RoadMapSslIO ssl_io_ctx = malloc( sizeof( struct roadmap_ssl_io_t ) );
   roadmap_check_allocated( ssl_io_ctx );

   ssl_io_ctx->ctx = ctx;
   ssl_io_ctx->ssl = ssl;
   ssl_io_ctx->s = s;
   ssl_io_ctx->data = data;
   ssl_io_ctx->on_connect = callback;

   if ( SSL_connect(ssl) == -1 )       /* perform the connection */
   {
      roadmap_log( ROADMAP_ERROR, "Error connecting ssl!!" );
      log_ssl();
      callback( s, data, ssl_io_ctx, err_net_failed );
      return 0;
   }

   callback( s, data, ssl_io_ctx, succeeded );

   return 1;
}

/*
 ******************************************************************************
 */

int roadmap_ssl_read (void *context, void *buffer, int buffer_size)
{
   int read;
   RoadMapSslIO ssl_io_ctx = (RoadMapSslIO) context;
   int received = 0;
   int err;
   do
   {
      read = SSL_read( ssl_io_ctx->ssl, buffer, buffer_size );

      buffer_size -= read;
      buffer += read;

      if ( ( err = SSL_get_error( ssl_io_ctx->ssl, read ) ) == SSL_ERROR_NONE )
      {
         received += read;
      }
      else
      {
         roadmap_log ( ROADMAP_ERROR, "Error (%d) in SSL_read: %s !!", err, ERR_error_string( err, NULL ) );
         received = 0;
         break;
      }

   } while ( SSL_pending( ssl_io_ctx->ssl ) && ( buffer_size > 0 ) );

   return received;
}

/*
 ******************************************************************************
 */

int roadmap_ssl_send ( RoadMapSslIO io, const void *data, int length)
{
   RoadMapSslIO ssl_io_ctx = io;
   int written = 0;
   int ret_code = SSL_write( ssl_io_ctx->ssl, data, length );

   switch( SSL_get_error( ssl_io_ctx->ssl, ret_code ) )
   {
             /* We wrote something*/
             case SSL_ERROR_NONE:
               written = ret_code;
               break;

             /* We would have blocked */
             case SSL_ERROR_WANT_WRITE:
                written = 0;
               break;
             case SSL_ERROR_WANT_READ:
                written = 0;
               break;

             /* Some other error */
             default:
                written = -1;
   }
   return written;
}

/*
 ******************************************************************************
 */
void roadmap_ssl_close ( void *context )
{
   RoadMapSslIO ssl_io_ctx = (RoadMapSslIO) context;
   if (context){
      SSL_free( ssl_io_ctx->ssl );                        /* release connection state */
      SSL_CTX_free( ssl_io_ctx->ctx );
   }
}

/*
 ******************************************************************************
 */
static const char* last_err_string( void )
{
   unsigned long last_error = ERR_peek_last_error();
   return ERR_error_string( last_error, NULL );
}
/*
 ******************************************************************************
 */
static void log_ssl( void )
{
  char buf[256];
  u_long err;
  while ( (err = ERR_get_error()) != 0 )
  {
     ERR_error_string_n(err, buf, sizeof(buf));
     roadmap_log( ROADMAP_ERROR, "SSL ERROR: *** %s", buf );
  }
}

